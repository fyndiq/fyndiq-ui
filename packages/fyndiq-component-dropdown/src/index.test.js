import React from 'react'
import { shallow, mount } from 'enzyme'

import Dropdown from './'

// Simulate events on <body>
// Usage : simulate.click('body') for a click event
//         simulate.keyup({ keyCode: 27 }) for an escape key press.
const simulate = {}
document.addEventListener = jest.fn((event, cb) => {
  simulate[event] = cb
})
document.removeEventListener = jest.fn((event, cb) => {
  if (simulate[event] === cb) {
    delete simulate[event]
  }
})

describe('fyndiq-component-dropdown', () => {
  test('should be displayed with minimum props', () => {
    expect(shallow(<Dropdown button="Button">Value</Dropdown>)).toMatchSnapshot()
  })

  test('should have a custom button', () => {
    expect(shallow(
      <Dropdown button={<div>Hello</div>}>
        Value
      </Dropdown>
    )).toMatchSnapshot()
  })

  test('should have a custom wrapper', () => {
    expect(shallow(
      <Dropdown button="B" noWrapperStyle opened>Content</Dropdown>
    ).find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should be opened when clicked on the button', () => {
    const component = mount(<Dropdown button="button">Value</Dropdown>)
    component.find('div > div').at(0).simulate('click')
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should be closed when clicking outside', () => {
    const component = mount(<Dropdown button="button">Value</Dropdown>)
    component.find('div > div').at(0).simulate('click')
    simulate.click('body')
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should be opened when hovered, if in hoverMode', () => {
    jest.useFakeTimers()
    const component = mount(<Dropdown button="button" hoverMode>Value</Dropdown>)
    component.simulate('mouseover')
    jest.runTimersToTime(200)
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
    component.simulate('mouseout')
    jest.runTimersToTime(100)
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should open the dropdown on button click in hoverMode', () => {
    jest.useFakeTimers()
    const component = mount(<Dropdown button="button" hoverMode>Value</Dropdown>)
    component.simulate('mouseover')
    jest.runTimersToTime(200)
    component.find('div > div').at(0).simulate('click')
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should have different dropdown positions', () => {
    expect(mount(
      <Dropdown button="B" position="bc" opened>Content</Dropdown>
    ).find('.dropdownWrapper')).toMatchSnapshot()

    expect(mount(
      <Dropdown button="B" position="tr" opened>Content</Dropdown>
    ).find('.dropdownWrapper')).toMatchSnapshot()

    expect(mount(
      <Dropdown button="B" position="tc" opened>Content</Dropdown>
    ).find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should have an arrow oriented to the top if dropdown is above', () => {
    expect(shallow(
      <Dropdown button="B" position="tr" opened>Content</Dropdown>
    ).find('Arrow')).toMatchSnapshot()
  })

  test('should be closable by pressing the Escape key', () => {
    const component = mount(<Dropdown button="button">Content</Dropdown>)
    component.find('div > div').at(0).simulate('click')
    simulate.keyup({ keyCode: 27 })
    expect(component.find('.dropdownWrapper')).toMatchSnapshot()
  })

  test('should remove eventListeners when unmounted', () => {
    const component = mount(<Dropdown button="button">Content</Dropdown>)
    component.unmount()
    expect(simulate).toMatchSnapshot()
  })
})
