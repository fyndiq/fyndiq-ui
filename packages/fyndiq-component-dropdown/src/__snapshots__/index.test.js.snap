// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fyndiq-component-dropdown should be closable by pressing the Escape key 1`] = `""`;

exports[`fyndiq-component-dropdown should be closed when clicking outside 1`] = `""`;

exports[`fyndiq-component-dropdown should be displayed with minimum props 1`] = `
<div
  className="wrapper"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    onClick={[Function]}
  >
    <Button
      disabled={false}
      horizontal={false}
      onClick={[Function]}
      pressed={false}
      size="s"
      type="normal"
    >
      Button
      <Arrow
        className=""
        orientation="down"
      />
    </Button>
  </div>
</div>
`;

exports[`fyndiq-component-dropdown should be opened when clicked on the button 1`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-bl
            "
  style={
    Object {
      "left": 0,
      "top": 5,
    }
  }
>
  Value
</div>
`;

exports[`fyndiq-component-dropdown should be opened when hovered, if in hoverMode 1`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-bl
            "
  style={
    Object {
      "left": 0,
      "top": 5,
    }
  }
>
  Value
</div>
`;

exports[`fyndiq-component-dropdown should be opened when hovered, if in hoverMode 2`] = `""`;

exports[`fyndiq-component-dropdown should have a custom button 1`] = `
<div
  className="wrapper"
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div
    onClick={[Function]}
  >
    <div>
      Hello
    </div>
  </div>
</div>
`;

exports[`fyndiq-component-dropdown should have a custom wrapper 1`] = `
<div
  className="
              dropdownWrapper
              
              position-bl
            "
  style={
    Object {
      "left": 0,
      "top": 0,
    }
  }
>
  Content
</div>
`;

exports[`fyndiq-component-dropdown should have an arrow oriented to the top if dropdown is above 1`] = `
<Arrow
  className=""
  orientation="up"
/>
`;

exports[`fyndiq-component-dropdown should have different dropdown positions 1`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-bc
            "
  style={
    Object {
      "left": 0,
      "top": 5,
    }
  }
>
  Content
</div>
`;

exports[`fyndiq-component-dropdown should have different dropdown positions 2`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-tr
            "
  style={
    Object {
      "left": 0,
      "top": -5,
    }
  }
>
  Content
</div>
`;

exports[`fyndiq-component-dropdown should have different dropdown positions 3`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-tc
            "
  style={
    Object {
      "left": 0,
      "top": -5,
    }
  }
>
  Content
</div>
`;

exports[`fyndiq-component-dropdown should open the dropdown on button click in hoverMode 1`] = `
<div
  className="
              dropdownWrapper
              dropdownDefault
              position-bl
            "
  style={
    Object {
      "left": 0,
      "top": 5,
    }
  }
>
  Value
</div>
`;

exports[`fyndiq-component-dropdown should remove eventListeners when unmounted 1`] = `
Object {
  "mouseout": [Function],
  "mouseover": [Function],
}
`;
